netsim/modules/bgp.py

```diff
@@ -178,38 +178,32 @@ def get_remote_ibgp_endpoint(n: Box) -> Box:
 * Other nodes need IBGP sessions with all RRs in the same AS
 """
 def build_ibgp_sessions(node: Box, sessions: Box, topology: Box) -> None:
-  rrlist = find_bgp_rr(node.bgp.get("as"),topology)
-  has_ibgp     = False                            # Assume we have no IBGP sessions (yet)
-
-  # If we don't have route reflectors, or if the current node is a route
-  # reflector, we need BGP sessions to all other nodes in the same AS
-  if not(rrlist) or node.bgp.get("rr",None):
-    for name,n in topology.nodes.items():
-      if "bgp" in n:
-        if n.bgp.get("as") == node.bgp.get("as") and n.name != node.name:
-          n_intf = get_remote_ibgp_endpoint(n)
-          neighbor_data = bgp_neighbor(n,n_intf,'ibgp',sessions,get_neighbor_rr(n))
-          if not neighbor_data is None:
-            if 'loopback' in node:
-              neighbor_data._source_intf = node.loopback
-            node.bgp.neighbors.append(neighbor_data)
-            has_ibgp = True
-
-  #
-  # The node is not a route reflector, and we have a non-empty RR list
-  # We need BGP sessions with the route reflectors
+  node_as  = node.bgp['as']                       # Set up variables that will come handy, starting with node AS
+  is_rr    = node.bgp.get("rr",None)              # Is this node an RR?
+  bgp_nhs  = node.bgp.get("next_hop_self",None)   # Do we have to set next hop on IBGP sessions?
+  rrlist = [] if is_rr else find_bgp_rr(node_as,topology)
+
+  if is_rr or not rrlist:                         # If the current node is RR or we have a full mesh
+    ibgp_ngb_list = [                             # ... we need IBGP sessions to all nodes in the AS
+      ngb for ngb in topology.nodes.values()      # ... but while building the node list
+        if ngb.name != node.name and              # ... skip current node
+           ngb.get('bgp.as',None) == node_as]     # ... and everyone not in the current AS (or not running BGP)
   else:
-    for n in rrlist:
-      if n.name != node.name:
-        n_intf = get_remote_ibgp_endpoint(n)
-        neighbor_data = bgp_neighbor(n,n_intf,'ibgp',sessions,get_neighbor_rr(n))
-        if not neighbor_data is None:
-          if 'loopback' in node:
-            neighbor_data._source_intf = node.loopback
-          node.bgp.neighbors.append(neighbor_data)
-          has_ibgp = True
-
-  if not has_ibgp:
+    ibgp_ngb_list = rrlist
+
+  for n in ibgp_ngb_list:
+    n_intf = get_remote_ibgp_endpoint(n)
+    neighbor_data = bgp_neighbor(n,n_intf,'ibgp',sessions,get_neighbor_rr(n))
+    if not neighbor_data is None:
+      if 'loopback' in node:
+        neighbor_data._source_intf = node.loopback
+      if bgp_nhs:
+        neighbor_data.next_hop_self = True
+      if is_rr and not 'rr' in neighbor_data:
+        neighbor_data.rr_client = True
+      node.bgp.neighbors.append(neighbor_data)
+
+  if not ibgp_ngb_list:
```

Comment:
This change leads to a warning about missing IBGP neighbors even though the topology specifies ```bgp.sessions.ipv4: [ ebgp ]``` to exclude any IBGP sessions

The difference seems to be that the ```has_ibgp``` flag wouldn't get set in that case, but the new ```ibgp_ngb_list``` contains nodes even though the bgp.sessions parameter says those sessions shouldn't get built